File: src/dada2.r
# DADA2 amplicon sequencing pipeline
#
# Processes paired- or single-end Illumina reads into ASVs and assigns
# taxonomy using a reference database (default: PR2). Options for output
# of bootstraps or counts only, as well as parameters for filterAndTrim(),
# dada(), mergePairs(), and assignTaxonomy().
#
# Usage:
#   Rscript dada2.r <config.yaml>
#
# Outputs (in workspace.root/Tables/):
#   seqtab_nochim.csv        - chimera-free ASV count table (sequences as rows)
#   asvs.fasta / asvs.csv    - ASV sequences with short identifiers (seq1, seq2 ...)
#   taxonomy.csv             - taxonomy assignments (optionally with bootstrap confidence values)
#   tax_counts.xlsx          - combined taxonomy + per-sample counts
#   pipeline_stats.csv       - read counts at each pipeline stage
#
# Notice:
#
# © 2026 Joshua Benjamin Jewell. All rights reserved.
#
# This module is licensed under the GNU Affero General Public License version 3 (AGPLv3).
#
# This work is based on the DADA2 tutorial by Benjamin J. Callahan, et al.,
# available at https://benjjneb.github.io/dada2/tutorial.html, with modification
# into a single module. The original material is licensed under the Creative
# Commons Attribution 4.0 International License (CC BY 4.0):
# https://creativecommons.org/licenses/by/4.0/.

source("./src/dada2_functions.r")

## Pipeline
main <- function() {
  # Load args
  args <- commandArgs(trailingOnly = TRUE)
  if (length(args) == 0) {
    stop("Usage: Rscript dada2_newer.r <config.yaml>")
  }
  yaml_path <- args[1]

  # Load config
  message("Loading and validating config: ", yaml_path)
  cfg     <- load_config(yaml_path)
  cfg     <- validate_config(cfg)
  verbose <- cfg$verbose %||% TRUE
  mode    <- cfg$file_patterns$mode %||% "paired"

  message("Setting up workspace: ", cfg$workspace$root)
  paths <- setup_workspace(cfg$workspace$root)

  # Find input files
  message("Discovering FASTQ files (mode: ", mode, ")")
  fastq <- find_fastq_files(
    cfg$workspace$input_dir,
    cfg$file_patterns$forward,
    cfg$file_patterns$reverse,
    mode
  )
  validate_sample_files(fastq$forward, fastq$reverse, mode)

  # In forward or reverse-only mode, sample names come from whichever side
  # is present; the split pattern in config determines what counts as a name.
  primary_files <- fastq$forward %||% fastq$reverse
  sample_names  <- extract_sample_names(
    primary_files,
    cfg$file_patterns$sample_name_split,
    cfg$file_patterns$sample_name_index
  )
  message("  Found ", length(sample_names), " samples")

  # Single-sample fallback:
  # dada() returns a bare dada object (not a list) when given a single file.
  # This breaks makeSequenceTable(), mergePairs(), and sapply() downstream.
  # Workaround: duplicate the file paths so the pipeline sees 2 samples, then
  # drop the duplicate row from all results before writing outputs.
  # ASV calls are unaffected because dada() processes each file independently.
  single_sample <- length(sample_names) == 1
  if (single_sample) {
    warning("Only 1 sample found. Duplicating it to work around dada() returning ",
            "a bare object for single-file input. The duplicate will be dropped ",
            "from all outputs.")
    fastq$forward <- rep(fastq$forward, 2)
    if (!is.null(fastq$reverse)) fastq$reverse <- rep(fastq$reverse, 2)
    sample_names <- c(sample_names, paste0(sample_names, "_dup"))
  }

  # Quality assessment (pre-filter)
  # Inspect quality_unfiltered.pdf to choose truncLen and maxEE values.
  message("Plotting unfiltered quality profiles")
  plot_quality_profiles(
    fastq$forward,
    fastq$reverse,
    file.path(paths$Figures, "quality_unfiltered.pdf")
  )

  # Filter and trim
  # Removes low-quality bases and reads. Reads that pass filtering are written
  # to Filtered/; empty samples (0 reads) are silently skipped by DADA2 later.
  message("Filtering and trimming reads")
  filtered <- make_filtered_paths(sample_names, paths$Filtered, mode)

  # For single-end modes, route the relevant files into the fwd slots and
  # pass NULL for rev so run_filter_trim() calls the correct filterAndTrim form.
  in_fwd      <- if (mode != "reverse") fastq$forward else fastq$reverse
  out_fwd     <- if (mode != "reverse") filtered$forward else filtered$reverse
  in_rev_arg  <- if (mode == "paired") fastq$reverse else NULL
  out_rev_arg <- if (mode == "paired") filtered$reverse else NULL

  filter_stats <- run_filter_trim(in_fwd, in_rev_arg, out_fwd, out_rev_arg,
                                  cfg$filter_trim, verbose)

  message("Plotting filtered quality profiles")
  plot_quality_profiles(
    filtered$forward,
    filtered$reverse,
    file.path(paths$Figures, "quality_filtered.pdf")
  )

  # Error model
  # set.seed() should be called before learnErrors()
  message("Learning error rates")
  set.seed(cfg$dada$seed %||% 123L)

  fwd_errors <- if (mode != "reverse") {
    learnErrors(filtered$forward,
                nbases = cfg$dada$nbases, MAX_CONSIST = cfg$dada$max_consist,
                verbose = verbose)
  } else NULL

  rev_errors <- if (mode != "forward") {
    learnErrors(filtered$reverse,
                nbases = cfg$dada$nbases, MAX_CONSIST = cfg$dada$max_consist,
                verbose = verbose)
  } else NULL

  # Inspect error_rates.pdf: the fitted line should follow the observed points.
  plot_error_rates(fwd_errors, rev_errors,
                   file.path(paths$Figures, "error_rates.pdf"))

  # Denoising
  # dada() applies the error model to resolve ASV's from sequencing errors.
  # pool = "pseudo" shares information across samples to improve detection of
  # rare variants.
  message("Denoising reads")
  dada_fwd <- if (!is.null(fwd_errors)) {
    dada(filtered$forward, err = fwd_errors, pool = cfg$dada$pool_method,
         verbose = verbose)
  } else NULL

  dada_rev <- if (!is.null(rev_errors)) {
    dada(filtered$reverse, err = rev_errors, pool = cfg$dada$pool_method,
         verbose = verbose)
  } else NULL

  # Merge and sequence table
  # Paired reads are joined at the overlapping region. Longer overlap and
  # lower maxMismatch = higher confidence merges but fewer total merges.
  message("Building sequence table")
  if (mode == "paired") {
    merged <- mergePairs(
      dada_fwd, filtered$forward,
      dada_rev, filtered$reverse,
      minOverlap   = cfg$merge$min_overlap,
      maxMismatch  = cfg$merge$max_mismatch,
      trimOverhang = cfg$merge$trim_overhang,
      verbose      = verbose
    )
    seq_table <- makeSequenceTable(merged)
  } else {
    merged    <- NULL
    seq_table <- makeSequenceTable(dada_fwd %||% dada_rev)
  }

  # Plot the full length distribution before any length filtering so that
  # off-target bands are visible when choosing band_size_min / band_size_max.
  plot_length_distribution(seq_table,
                           file.path(paths$Figures, "length_distribution.pdf"))

  # Length filtering: retains only ASVs matching the expected amplicon size.
  # Set band_size_min / band_size_max to null in config to skip this step.
  band_min <- cfg$asv$band_size_min
  band_max <- cfg$asv$band_size_max
  if (!is.null(band_min) && !is.null(band_max)) {
    message("  Filtering by length: ", band_min, "-", band_max, " bp")
    seq_table <- filter_by_length(seq_table, band_min, band_max)
    # Second plot confirms off-target lengths were successfully removed.
    plot_length_distribution(seq_table,
                             file.path(paths$Figures, "length_distribution_filtered.pdf"))
  }

  # Chimera removal
  # "consensus" removes sequences identified as chimeric in the majority
  # of samples in which they appear.
  message("Removing chimeras")
  seq_table_nochim <- removeBimeraDenovo(seq_table, method = cfg$asv$denovo_method,
                                         verbose = verbose)

  nochim_pct <- sum(seq_table_nochim) / sum(seq_table) * 100
  message("  Chimeric reads removed: ", round(100 - nochim_pct, 2),
          "% | Retained: ", round(nochim_pct, 2), "%")

  # Drop duplicate row (if using single-sample fallback)
  if (single_sample) {
    filter_stats     <- filter_stats[1, , drop = FALSE]
    if (!is.null(dada_fwd)) dada_fwd <- dada_fwd[1]
    if (!is.null(dada_rev)) dada_rev <- dada_rev[1]
    if (!is.null(merged))   merged   <- merged[1]
    seq_table_nochim <- seq_table_nochim[1, , drop = FALSE]
    sample_names     <- sample_names[1]
  }

  # Pipeline stats
  message("Computing pipeline stats")
  stats <- compute_pipeline_stats(filter_stats, dada_fwd, dada_rev, merged,
                                  seq_table_nochim, sample_names, mode)
  write.csv(stats, file.path(paths$Tables, "pipeline_stats.csv"), quote = FALSE)
  if (verbose) print(stats)

  # Write core outputs
  write_seq_table(seq_table_nochim, paths$Tables,
                  cfg$output$seq_table_prefix %||% "seqtab_nochim")

  message("Writing output tables")
  # write_fasta() returns index (SeqName - Sequence), which is the join key
  # used in write_taxa_table() and write_combined_table() below.
  index <- write_fasta(seq_table_nochim, paths$Tables,
                       cfg$output$fasta_prefix %||% "asvs")

  # combined_input() defaults to index (SeqName + Sequence + counts only),
  # which is the alternative output mode. It is replaced with the full
  # taxa_df when taxonomy is run and combined_mode = "regular".
  combined_input <- index

  # Taxonomy
  if (!(cfg$taxonomy$skip %||% FALSE)) {
    message("Assigning taxonomy")
    db_path     <- fetch_taxonomy_db(cfg$taxonomy$uri, paths$Taxonomy)
    taxa_result <- run_assign_taxonomy(seq_table_nochim, db_path,
                                       cfg$taxonomy, verbose)
    taxa_df <- write_taxa_table(
      taxa_result$tax, taxa_result$boot, index,
      paths$Tables,
      cfg$output$taxa_prefix %||% "taxonomy",
      cfg$output$bootstraps %||% "combined"
    )
    if ((cfg$output$combined_mode %||% "regular") == "regular") {
      combined_input <- taxa_df
    }
  } else {
    message("  Skipping taxonomy (taxonomy.skip = true)")
  }

  # Checkpoint saved after taxonomy so the full environment (including taxa)
  # can be restored with load() without re-running the pipeline.
  checkpoint <- file.path(paths$Analysis, "checkpoint.RData")
  save.image(checkpoint)
  message("  Checkpoint saved: ", checkpoint)

  write_combined_table(combined_input, seq_table_nochim, paths$Tables,
                       cfg$output$combined_filename %||% "tax_counts.xlsx")

  message("Pipeline complete.")
  message("Outputs:")
  message("  ", file.path(paths$Tables, paste0(cfg$output$seq_table_prefix %||% "seqtab_nochim", ".csv")))
  message("  ", file.path(paths$Tables, paste0(cfg$output$fasta_prefix %||% "asvs", ".fasta")))
  message("  ", file.path(paths$Tables, paste0(cfg$output$fasta_prefix %||% "asvs", ".csv")))
  message("  ", file.path(paths$Tables, paste0(cfg$output$taxa_prefix %||% "taxonomy", ".csv")))
  message("  ", file.path(paths$Tables, cfg$output$combined_filename %||% "tax_counts.xlsx"))
  message("  ", file.path(paths$Tables, "pipeline_stats.csv"))
  message("  ", checkpoint)
}

# Run pipeline
if (sys.nframe() == 0L) main()

File: src/pipeline/dada2.jl
module DADA2

# DADA2 amplicon sequencing pipeline - Julia orchestrator
#
# Stages and their Web UI page groupings:
#
#   QC page:        prefilter_qc, filter_trim
#   Denoising page: learn_errors, denoise, filter_length, chimera_removal
#   Taxonomy page:  assign_taxonomy
#
# Each stage is independently callable and saves a checkpoint so the pipeline
# can be resumed or individual stages re-run across Julia sessions.
#
# Call dada2() to run all stages in sequence without stopping.
#
# Notice:
#
# © 2026 Joshua Benjamin Jewell. All rights reserved.
#
# This module is licensed under the GNU Affero General Public License version 3 (AGPLv3).
#
# This work is based on the DADA2 tutorial by Benjamin J. Callahan, et al.,
# available at https://benjjneb.github.io/dada2/tutorial.html, with modification
# into a single module. The original material is licensed under the Creative
# Commons Attribution 4.0 International License (CC BY 4.0):
# https://creativecommons.org/licenses/by/4.0/.

export dada2, dada2_denoise, dada2_classify,
       prefilter_qc, filter_trim, learn_errors, denoise,
       filter_length, chimera_removal, assign_taxonomy

    import Downloads
    using Logging, RCall, YAML
    using ..PipelineTypes
    using ..PipelineLog
    using ..Databases
    using ..Config

    include("dada2/context.jl")   # shared helpers and _pipeline_context
    include("dada2/qc.jl")        # prefilter_qc, filter_trim
    include("dada2/denoise.jl")   # learn_errors, denoise, filter_length
    include("dada2/chimera.jl")   # chimera_removal
    include("dada2/taxonomy.jl")  # assign_taxonomy

    """
        dada2(config_path; progress)

    Run the complete DADA2 pipeline from raw reads to a taxonomy-annotated ASV
    table, calling all stages in sequence:

        prefilter_qc -> filter_trim -> learn_errors ->
        denoise -> filter_length -> chimera_removal -> assign_taxonomy

    For interactive use - reviewing intermediate outputs or adjusting parameters
    between steps - call the individual stage functions directly instead.

    ## Outputs
    Written to `workspace.root/Tables/`:
    - `seqtab_nochim.csv`              - chimera-free ASV count table
    - `asvs.fasta` / `asvs.csv`       - ASV sequences with short identifiers
    - `taxonomy.csv`                   - taxonomy assignments
    - `taxonomy_bootstraps.csv`        - bootstrap confidence values
    - `taxonomy_combined.csv`          - taxonomy + bootstraps combined
    - `tax_counts.csv`                 - taxonomy + per-sample counts
    - `asv_counts.csv`                 - ASV sequences + per-sample counts (no taxonomy)
    - `pipeline_stats.csv`             - read counts at each pipeline stage

    Written to `workspace.root/Checkpoints/`:
    - `ckpt_filter.RData`  - filter_stats
    - `ckpt_errors.RData`  - fwd_errors, rev_errors
    - `ckpt_denoise.RData` - dada objects, merged, unfiltered seq_table
    - `ckpt_length.RData`  - dada objects, merged, length-filtered seq_table
    - `ckpt_chimera.RData` - seq_table_nochim, index
    - `checkpoint.RData`   - final R environment snapshot
    """
    function dada2(config_path::String; progress=nothing, input_dir=nothing, workspace_root=nothing, taxonomy_db=nothing)
        R"rm(list=ls())"
        prefilter_qc(config_path; progress, input_dir, workspace_root)
        filter_trim(config_path; progress, input_dir, workspace_root); R"gc()"
        learn_errors(config_path; progress, input_dir, workspace_root); R"gc()"
        denoise(config_path; progress, input_dir, workspace_root); R"gc()"
        filter_length(config_path; progress, input_dir, workspace_root); R"gc()"
        chimera_removal(config_path; progress, input_dir, workspace_root); R"gc()"
        assign_taxonomy(config_path; progress, input_dir, workspace_root, taxonomy_db)
    end

    """
        dada2_denoise(project, trimmed; progress) -> DenoisedASVs

    Run all DADA2 stages up to and including chimera removal. Returns a
    `DenoisedASVs` wire type that can be passed to `dada2_classify()`.

    Skips automatically if `ckpt_chimera.RData` is newer than both `dada2.yml`
    and the trimmed reads. Individual stage skip guards apply for partial runs.
    """
    function dada2_denoise(project::ProjectCtx, trimmed::TrimmedReads; progress=nothing)
        config_path    = write_run_config(project)
        workspace_root = joinpath(project.dir, "dada2")
        chimera_ckpt   = joinpath(workspace_root, "Checkpoints", "ckpt_chimera.RData")

        trimmed_files = filter(f -> endswith(f, "_trimmed.fastq.gz"), readdir(trimmed.dir))
        trimmed_mtime = isempty(trimmed_files) ? 0.0 :
                        maximum(mtime(joinpath(trimmed.dir, f)) for f in trimmed_files)

        if isfile(chimera_ckpt) &&
           mtime(chimera_ckpt) > mtime(config_path) &&
           mtime(chimera_ckpt) > trimmed_mtime
            @info "[$(basename(dirname(workspace_root)))] DADA2: Skipping denoise - ckpt_chimera.RData up to date"
        else
            R"rm(list=ls())"
            prefilter_qc(config_path;    progress, input_dir=trimmed.dir, workspace_root)
            filter_trim(config_path;     progress, input_dir=trimmed.dir, workspace_root); R"gc()"
            learn_errors(config_path;    progress, input_dir=trimmed.dir, workspace_root); R"gc()"
            denoise(config_path;         progress, input_dir=trimmed.dir, workspace_root); R"gc()"
            filter_length(config_path;   progress, input_dir=trimmed.dir, workspace_root); R"gc()"
            chimera_removal(config_path; progress, input_dir=trimmed.dir, workspace_root); R"gc()"
            pipeline_log(project, "DADA2 denoising complete")
        end

        return DenoisedASVs(chimera_ckpt, config_path, workspace_root, trimmed.dir)
    end

    """
        dada2_classify(project, denoised; taxonomy_db, progress) -> ASVResult

    Run taxonomy assignment on a `DenoisedASVs` result and return an `ASVResult`.
    Respects the mtime skip guard inside `assign_taxonomy()`.
    """
    function dada2_classify(project::ProjectCtx, denoised::DenoisedASVs;
                            taxonomy_db=nothing, progress=nothing, skip_taxonomy=false)
        config_path    = denoised.config_path
        workspace_root = denoised.workspace_root
        cfg        = get(YAML.load_file(config_path), "dada2", Dict())
        out_cfg    = get(cfg, "output", Dict())
        tables_dir = joinpath(workspace_root, "Tables")
        result = ASVResult(
            joinpath(tables_dir, get(out_cfg, "fasta_prefix",     "asvs")          * ".fasta"),
            joinpath(tables_dir, get(out_cfg, "seq_table_prefix", "seqtab_nochim") * ".csv"),
            joinpath(tables_dir, get(out_cfg, "taxa_prefix",      "taxonomy")      * ".csv")
        )
        assign_taxonomy(config_path;
                        progress,
                        input_dir      = denoised.input_dir,
                        workspace_root,
                        taxonomy_db,
                        skip_classification = skip_taxonomy)
        pipeline_log(project, "DADA2 taxonomy complete")
        R"rm(list=ls()); gc()"
        for f in (result.fasta, result.count_table, result.taxonomy)
            isfile(f) && log_written(project, f)
        end
        return result
    end

    function dada2(project::ProjectCtx, trimmed::TrimmedReads;
                   taxonomy_db=nothing, progress=nothing, skip_taxonomy=false)
        config_path    = write_run_config(project)
        workspace_root = joinpath(project.dir, "dada2")
        cfg        = get(YAML.load_file(config_path), "dada2", Dict())
        out_cfg    = get(cfg, "output", Dict())
        tables_dir = joinpath(workspace_root, "Tables")
        result = ASVResult(
            joinpath(tables_dir, get(out_cfg, "fasta_prefix",     "asvs")          * ".fasta"),
            joinpath(tables_dir, get(out_cfg, "seq_table_prefix", "seqtab_nochim") * ".csv"),
            joinpath(tables_dir, get(out_cfg, "taxa_prefix",      "taxonomy")      * ".csv")
        )
        trimmed_files = filter(f -> endswith(f, "_trimmed.fastq.gz"), readdir(trimmed.dir))
        trimmed_mtime = isempty(trimmed_files) ? 0.0 :
                        maximum(mtime(joinpath(trimmed.dir, f)) for f in trimmed_files)
        if all(isfile, (result.fasta, result.count_table, result.taxonomy)) &&
           all(f -> mtime(f) > mtime(config_path), (result.fasta, result.count_table, result.taxonomy)) &&
           all(f -> mtime(f) > trimmed_mtime, (result.fasta, result.count_table, result.taxonomy))
            @info "[$(basename(project.dir))] DADA2: Skipping - outputs up to date in $tables_dir"
            return result
        end
        @info "[$(basename(project.dir))] DADA2: Starting pipeline"
        pipeline_log(project, "DADA2 pipeline started")
        denoised = dada2_denoise(project, trimmed; progress)
        result = dada2_classify(project, denoised; taxonomy_db, progress, skip_taxonomy)
        pipeline_log(project, "DADA2 complete")
        return result
    end

end
